src/server.rs

```diff
@@ -0,0 +1,221 @@
+use crate::config::CONFIG;
+use bytes::{BufMut, BytesMut};
+use std::{borrow::Cow, fmt::Debug, str};
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    net::{TcpListener, TcpStream, ToSocketAddrs},
+};
+
+#[derive(Debug)]
+pub enum Command {
+    Handshake,          // HI
+    ClientVersion,      // ID
+    KeepAlive,          // CH
+    AskListLengths,     // askchaa
+    AskListCharacters,  // askchar
+    CharacterList,      // AN
+    EvidenceList,       // AE
+    MusicList,          // AM
+    AO2CharacterList,   // RC
+    AO2Ready,           // RD
+    SelectCharacter,    // CC
+    ICMessage,          // MS
+    OOCMessage,         // CT
+    PlaySong,           // MC
+    WTCEButtons,        // RT
+    SetCasePreferences, // SETCASE
+    CaseAnnounce,       // CASEA
+    Penalties,          // HP
+    AddEvidence,        // PE
+    DeleteEvidence,     // DE
+    EditEvidence,       // EE
+    CallModButton,      // ZZ
+    KickWithGuard,      // opKICK
+    BanWithGuard,       // opBAN
+}
+
+impl FromStr for Command {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        use Command::*;
+        match s {
+            "HI" => Ok(Handshake),
+            "ID" => Ok(ClientVersion),
+            "CH" => Ok(KeepAlive),
+            "askchaa" => Ok(AskListLengths),
+            "askchar" => Ok(AskListCharacters),
+            "AN" => Ok(CharacterList),
+            "AE" => Ok(EvidenceList),
+            "AM" => Ok(MusicList),
+            "RC" => Ok(AO2CharacterList),
+            "RD" => Ok(AO2Ready),
+            "CC" => Ok(SelectCharacter),
+            "MS" => Ok(ICMessage),
+            "CT" => Ok(OOCMessage),
+            "MC" => Ok(PlaySong),
+            "RT" => Ok(WTCEButtons),
+            "SETCASE" => Ok(SetCasePreferences),
+            "CASEA" => Ok(CaseAnnounce),
+            "HP" => Ok(Penalties),
+            "PE" => Ok(AddEvidence),
+            "DE" => Ok(DeleteEvidence),
+            "EE" => Ok(EditEvidence),
+            "ZZ" => Ok(CallModButton),
+            "opKICK" => Ok(KickWithGuard),
+            "opBAN" => Ok(BanWithGuard),
+            _ => anyhow::bail!("Invalid command!"),
+        }
+    }
+}
+
+enum DecodeState {
+    Command,
+    Arguments,
+}
+
+use serde::export::Formatter;
+use std::{
+    io::{BufRead, Cursor, Seek, SeekFrom},
+    str::FromStr,
+};
+use tokio::{
+    io::{AsyncBufReadExt, Error, ErrorKind},
+    stream::StreamExt,
+};
+use tokio_util::codec::{Decoder, Encoder, FramedRead};
+
+#[derive(Debug)]
+pub struct AOMessage {
+    pub command: Command,
+    pub args: Vec<String>,
+}
```

Comment:
1. Many allocations for String.
2. Not type-safe. I think it would be better if we added arguments to Command enum.
3. I think it would be better if we create trait Command with function from_protocol, and use it trait instead of struct - it will more flexible.